Add support for LayeredHitSounds skin option

Fire937 opened the pull request:
Closes  #8595

The current behaviour of osu!lazer is to add the hitnormal sample on top of the other samples for every single skin/beatmap/ruleset/hitobject, even in osu!mania, that is what causes the issue. Moreover osu!stable uses a LayeredHitSounds skin property that tells osu! whether it should add the hitnormal sample artificially or not based on the skin used (defaults to yes). LayeredHitSounds is also ignored when in osu!mania mode: osu!stable never adds the hitnormal sample artificially in osu!mania mode.

Currently the `LayeredHitSounds` property is included in `LegacySkinConfiguration,` however i am not sure about whether it belongs here or not. Maybe it should be moved to `SkinInfo`.
A ruleset can override the value of `AllowsLayeredHitSounds` in its `ConvertHitObjectParser` to disable the `LayeredHitSounds` property.

- [x] Depends on #8649

Fire937 committed:
Make Beatmap Decoder consistent with osu-stable

The 5th argument of a timing point was not read properly. When equal to
0 the sample used should be osu!'s default hitsound. This commit
resolves this problem by adding an 'IsCustom' attribute to
HitSampleInfo.

Fire937 committed:
Format code and rewrite innacurate comments

Fire937 committed:
Format code

Fire937 committed:
Fix code style

Fire937 committed:
Merge branch 'master' into audio-rework

Fire937 committed:
Move the verification of a sample's customness to LegacySkin

As suggested by @bdach it makes much more sense to test the IsCustom
attribute to LegacySkin, that way SkinnableSound doesn't need to know
how the skin is supposed to behave.

Fire937 committed:
Merge branch 'audio-rework' of github.com:Fire937/osu into audio-rework

Fire937 committed:
Add support for LayeredHitSounds skin option

The LayeredHitSounds option if set to 1 will force hitnormal to be
played along all the other hitsounds on each hitobject.

This option is disabled in osu:mania. In order to implement this
behaviour, an AllowsLayeredHitSounds attribute was added to the
ConvertHitObjectParser so that it knows if it has to artificially add
the hitnormal sample or not. LegacySkin can then verify if a hitsample
was artificially added thanks to HitSampleInfo's IsLayered attribute. It
can then decide to return the hitsample or not based on the value of
LayeredHitSounds.

Fire937 committed:
Merge branch 'audio-rework' of github.com:Fire937/osu into audio-rework-mania

These two branches are tightly related for the most part because of the
changes to LegacySkin:GetSample.

Fire937 committed:
Move LayeredHitSounds atttribute

LayeredHitSounds was previously in SkinInfo but that caused some issues
so it was moved to LegacySkinConfiguration.

Fire937 committed:
Make doc more consistent and enhance code style

Fire937 committed:
Add LayeredHitSounds setting to LegacySetting

Fire937 commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -304,17 +304,28 @@ public override Texture GetTexture(string componentName)
-            foreach (var lookup in sampleInfo.LookupNames)
+            var hsi = sampleInfo as HitSampleInfo;
+
+            if (hsi != null && hsi.IsLayered && !Configuration.LayeredHitSounds)
```
I agree. I changed it.

iiSaLMaN commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -124,6 +124,9 @@ protected override void Dispose(bool isDisposing)
                     {
                         case LegacySkinConfiguration.LegacySetting.Version:
                             return SkinUtils.As<TValue>(new Bindable<decimal>(Configuration.LegacyVersion ?? LegacySkinConfiguration.LATEST_VERSION));
+
+                        case LegacySkinConfiguration.LegacySetting.LayeredHitSounds:
+                            return SkinUtils.As<TValue>(new Bindable<bool>(Configuration.LayeredHitSounds ?? true));
```
This is actually redundant, to be precise on what you should do, `LegacySkinDecoder` simply takes all lines that're not handled specially as keys and values added to `LegacySkinConfiguration.ConfigDictionary`, so you would only need the enum member `LayeredHitSounds` to retrieve the value from the dictionary with, i.e. remove the property from `LegacySkinConfiguration` and remove the case added in `LegacySkinDecoder`.

Fire937 replied:
I see, thank you for your explanation I definitely misunderstood how the `SkinConfiguration` class works. I removed the property from `LegacySkinConfiguration` as you suggested and now access the property by string:
`GetConfig<string, bool>("LayeredHitSounds")`
Is that the proper way to do it?

iiSaLMaN replied:
Almost there! the lookup should've been `LegacySetting.LayeredHitSounds` as the name of the enum member `LayeredHitSounds` will then be converted to a string to use as the lookup key in the dictionary, hence why the lookup type is almost always an enum, that's because it makes life easier :P, we definitely don't want to search what was the name of a config value lookup everytime we want to retrieve one.

Fire937 replied:
That's what I thought at first but then in `GetConfig` there is no default case for `LegacySkinConfiguration.LegacySetting` meaning it only tests for `LegacySkinConfiguration.LegacySetting.Version` and breaks, thus returning null when trying to retrieve the `LayeredHitSounds` property.

Maybe the case in `GetConfig` should be kept or a default case that translates the enum to a string could be added. However that would be exactly what the default case does when it is none of the enums that have specific cases thus inducing code redundancy.

iiSaLMaN replied:
sounds like the version case should be:
```csharp
case LegacySkinConfiguration.LegacySetting legacySetting when legacySetting == LegacySkinConfiguration.LegacySetting.Version:
    // retrieve version.
```

Fire937 committed:
Remove LayeredHitSounds from Configuration

The use of a LayeredHitSounds property in Configuration was redundant.
The value of the LayeredHitSounds property could already be retrieved
from SkinConfiguration:ConfigDictionary.

bdach commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -304,17 +304,28 @@ public override Texture GetTexture(string componentName)
+                // The layered hitsample is ignored if the skin disabled LayeredHitSounds
+                return null;
+
+            if (hsi == null || hsi.IsCustom)
```
this should go in the PR this one depends on

Fire937 committed:
Add LayeredHitSounds to LegacySetting

The LayeredHitSounds property can now be accessed via the LegacySetting
enum rather than by string.

bdach commented on osu.Game/Audio/HitSampleInfo.cs:
```diff
@@ -37,6 +37,16 @@ public class HitSampleInfo : ISampleInfo
         public int Volume { get; set; }
 
+        /// <summary>
+        /// Whether the sample was artificially added or if it was a beatmap specification.
```
This xmldoc feels like a mental shortcut (what connection does "layering" have with "artificially adding a sample"? I wouldn't understand what it means without the PR description) I'd say it needs to mention explicitly that `IsLayered` is a normal hitsample layered over the beatmap one.

Fire937 replied:
That's true, what about: Whether the sample is layered over the beatmap ones or not.

bdach replied:
I'd go for the very explicit
```suggestion
        /// True if the sample is a <c>hitnormal</c> sample added artificially due to the skin having
        /// the <see cref="osu.Game.Skinning.LegacySkinConfiguration.LegacySetting.LayeredHitSounds"/> option set.
```

But then again adding *that* kind of shows that if possible, `IsLayered` shouldn't really be a flag at all as it is out of place in the general `HitSampleInfo` (it's something that has explicitly to do with legacy skins).

Fire937 replied:
In that regard I was thinking about making the `LegacyHitSampleInfo` class public and including the `IsLayered` property in it. Would that work or is it a bad idea? 
Incidentally `IsCustom` could also be included in it.

bdach commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -304,17 +304,28 @@ public override Texture GetTexture(string componentName)
+            var hsi = sampleInfo as HitSampleInfo;
+
+            if (hsi?.IsLayered == true && GetConfig<string, bool>("LayeredHitSounds")?.Value == false)
+                // The layered hitsample is ignored if the skin disabled LayeredHitSounds
```
Kind of a redundant comment.

bdach commented on osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:
```diff
@@ -32,6 +32,8 @@ public abstract class ConvertHitObjectParser : HitObjectParser
 
         protected bool FirstObject { get; private set; } = true;
 
+        protected bool AllowsLayeredHitSounds { get; set; } = true;
```
Usually we use infinitives for these sorts of flags (so would be `AllowLayeredHitSounds`)

Fire937 replied:
My bad, i am not extremely familiar yet with the conventions used in this repo. I will change it.

bdach commented on osu.Game/Rulesets/Objects/Legacy/Mania/ConvertHitObjectParser.cs:
```diff
@@ -15,6 +15,8 @@ public class ConvertHitObjectParser : Legacy.ConvertHitObjectParser
         public ConvertHitObjectParser(double offset, int formatVersion)
             : base(offset, formatVersion)
         {
+            // LayeredHitSounds is disabled in mania
```
Also sort of obvious.

bdach commented on osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:
```diff
@@ -349,16 +351,19 @@ private List<HitSampleInfo> convertSoundType(LegacyHitSoundType type, SampleBank
-                }
-            };
+                    CustomSampleBank = bankInfo.CustomSampleBank,
+                    IsLayered = !type.HasFlag(LegacyHitSoundType.Normal)
```
I feel like this is backwards. Shouldn't this just use `AllowsLayeredHitSounds`? If both the flag is set and `AllowsLayeredHitSounds` is `true`, won't this wrongly consider the sample non-layered?

Fire937 replied:
The thing is that the beatmap could actually explicitly indicate that the sample to be played (among others) should be hitnormal. In this case, whether `AllowLayeredHitSounds` is `true` or `false` should not affect the fact that the sample is not layered. That means it should always be played even in osu!mania. That's why this problem doesn't appear in other beatmap as they use hitnormal as opposed to this map that uses hitfinish and the others but not hitnormal.
I'm actually concerned about the naming of this flag, maybe `IsForced` would make more sense?

smoogipoo replied:
With application of the above comments, the outer condition can once again be removed and this can remain here.

bdach reviewed (changes requested):
I really dislike how complicated `LegacySkin.GetSample()` is starting to be. I'd say it needs unit tests at the least, and a refactoring on top to decrease complexity if possible.

Fire937 committed:
Add test for LayeredHitSounds

This test makes sure the layered hitsounds are handled properly by
LegacySkin:GetSample.
- If the LayeredHitSounds skin property is set to true, the sample
  should be played whether it is layered or not.
- If the LayeredHitSounds skin property is set to false, the sample
  should be played only if it is not layered.

Fire937 committed:
Format code

Fire937 committed:
Add test for handling non custom samples

Makes sure that LegacySkin.GetSample() returns the right samples
according to the IsCustom flag.

Fire937 committed:
Fix typo

Fire937 committed:
Merge branch 'audio-rework-mania' of github.com:Fire937/osu into audio-rework-mania

Fire937 committed:
Merge branch 'master' of https://github.com/ppy/osu into audio-rework-mania

smoogipoo commented on osu.Game/Beatmaps/Formats/LegacyDecoder.cs:
```diff
@@ -171,6 +171,9 @@ public override HitSampleInfo ApplyTo(HitSampleInfo hitSampleInfo)
                 if (string.IsNullOrEmpty(baseInfo.Suffix) && CustomSampleBank > 1)
                     baseInfo.Suffix = CustomSampleBank.ToString();
 
+                // 0 means osu!'s default hitsounds
+                baseInfo.IsCustom = CustomSampleBank != 0;
+
```
Same here, can be removed

smoogipoo commented on osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:
```diff
@@ -32,6 +32,8 @@ public abstract class ConvertHitObjectParser : HitObjectParser
 
         protected bool FirstObject { get; private set; } = true;
 
+        protected bool AllowLayeredHitSounds { get; set; } = true;
```
This cannot exist in the parser - it should only exist within the legacy skin itself.

The mania override can be done inside `LegacyManiaSkinTransformer`.

That is to say, layered hitsounds should _always_ be added.

smoogipoo commented on osu.Game/Audio/HitSampleInfo.cs:
```diff
@@ -37,6 +37,17 @@ public class HitSampleInfo : ISampleInfo
+        /// True if the sample is a <c>hitnormal</c> sample added artificially due to the skin having
+        /// the <see cref="osu.Game.Skinning.LegacySkinConfiguration.LegacySetting.LayeredHitSounds"/> option set.
+        /// </summary>
+        public bool IsLayered { get; set; }
```
This must only exist within `ConvertHitObjectParser.LegacyHitSampleInfo`.

smoogipoo commented on osu.Game/Rulesets/Objects/Legacy/Mania/ConvertHitObjectParser.cs:
```diff
@@ -15,6 +15,7 @@ public class ConvertHitObjectParser : Legacy.ConvertHitObjectParser
         public ConvertHitObjectParser(double offset, int formatVersion)
             : base(offset, formatVersion)
         {
+            AllowLayeredHitSounds = false;
```
See other comments.

smoogipoo commented on osu.Game.Tests/Skins/TestSceneLegacySkin.cs:
```diff
@@ -0,0 +1,105 @@
+namespace osu.Game.Tests.Skins
+{
+    [HeadlessTest]
+    public class TestSceneLegacySkin : OsuTestScene
```
This testing is inadequate because it doesn't test the result of gameplay loading. You want to add a test to `TestSceneHitObjectSamples`.

I'm fairly certain the implementation in this PR (inside `LegacySkin`) will fail, since `SkinnableSound` will fall back to the default lazer sample store if `null` is returned.

smoogipoo commented on osu.Game/Skinning/LegacySkin.cs:
```diff
@@ -311,17 +305,27 @@ public override Texture GetTexture(string componentName)
-            foreach (var lookup in sampleInfo.LookupNames)
+            var hsi = sampleInfo as HitSampleInfo;
+
+            if (hsi?.IsLayered == true && GetConfig<LegacySkinConfiguration.LegacySetting, bool>(LegacySkinConfiguration.LegacySetting.LayeredHitSounds)?.Value == false)
```
This isn't being read from the skin configuration. You need to add a value to the `LegacySetting` switch above.

peppy commented:
@Fire937 just looping back – are you intending to apply the remaining review?

Fire937 commented:
@peppy yes I will try and commit in not too long.

bdach commented:
Seeing that there stlll has been no changes here in a while, I would be looking to supersede this PR over the coming days. I don't usually do that but this one would take too much work to fix up - I would basically have to revert all of the changes here, and even then there are major issues left unaddressed (such as how to make the sample fallback logic stick to a silent sample if layered hit sounds are to be disabled).
